server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -1532,6 +1532,22 @@ public CanMatchShardResponse canMatch(ShardSearchRequest request) throws IOExcep
 
     private CanMatchShardResponse canMatch(ShardSearchRequest request, boolean checkRefreshPending) throws IOException {
         assert request.searchType() == SearchType.QUERY_THEN_FETCH : "unexpected search type: " + request.searchType();
+        if (request.source() != null && (request.source().query() instanceof MatchNoneQueryBuilder) == false) {
+            final SearchExecutionContext searchExecutionContext = new SearchExecutionContext(
+                indicesService.indexServiceSafe(request.shardId().getIndex())
```

Comment:
We should reuse the `indexService` below, using `indexServiceSafe` might not work for PIT if the index was deleted after the creation. Can you move this code [here](https://github.com/elastic/elasticsearch/pull/96161/files#diff-44f316dd7fee3be3afb1c7bca7b3b6fb86a4eac04bd0f776e7f3ae52ad7a541fL1575)?

Reply:
The idea of doing it at the beginning was to do a early exit and avoid doing everything that happens after...

Reply:
These operations are cheap (no IO) so it shouldn't be an issue and we want to handle all cases (PIT and normal search)?

Reply:
I think we need to do this rewrite before we invoked `IndexShard#acquireSearcher(...)` since this marks the shard as search active again (and would allow scheduled refreshes to really be executed). So maybe we should keep the rewrite in the place it was?

Reply:
If we do it late shards go into 'search active state'. `acquireSearcher` does `markSearcherAccessed `. WHich makes the shard search active which will cause waiting for refresh...which we want to avoid.

Reply:
I see and the PIT case sets the refresh pending to false anyway so no need to change that. Thanks for spotting.